File: programmers/탐욕법(Greedy)/체육복(Level1).py
'''
    Code written on Aug 19, 2020
    출처: 프로그래머스 코딩 테스트 연습, https://programmers.co.kr/learn/challenges
    탐욕법(Greedy): 체육복(Level1)
    link: https://programmers.co.kr/learn/courses/30/lessons/42862
'''
def solution(n, lost, reserve):
    intersection = set(lost) & set(reserve)
    lost = set(lost) - intersection
    reserve = set(reserve) - intersection
    answer = n - len(lost)
    for without in sorted(lost):
        if without-1 in reserve:
            reserve.discard(without-1)
            answer += 1
        elif without+1 in reserve:
            reserve.discard(without+1)
            answer += 1
        if not reserve:
            break
    return answer

File: programmers/탐욕법(Greedy)/체육복(Level1).java
/*
    Code written on Sep 23, 2020
    출처: 프로그래머스 코딩 테스트 연습, https://programmers.co.kr/learn/challenges
    탐욕법(Greedy): 체육복(Level1)
    link: https://programmers.co.kr/learn/courses/30/lessons/42862
*/
import java.util.Arrays;
import java.util.HashSet;

class Solution {
    public int solution(int n, int[] lost, int[] reserve) {
        HashSet<Integer> lostSet = new HashSet<Integer>();
        HashSet<Integer> reserveSet = new HashSet<Integer>();
        Arrays.sort(lost);

        for (int num: lost)
            lostSet.add(num);

        for (int num: reserve) {
            if (lostSet.contains(num))
                lostSet.remove(num);
            else
                reserveSet.add(num);
        }

        for (int num: lost) {
            if (lostSet.contains(num)) {
                if (reserveSet.contains(num-1)) {
                    lostSet.remove(num);
                    reserveSet.remove(num-1);
                } else if (reserveSet.contains(num+1)) {
                    lostSet.remove(num);
                    reserveSet.remove(num+1);
                }
            }
        }

        return n - lostSet.size();
    }
}
